src/policy/immix/vo_bit_helper.rs

```diff
@@ -0,0 +1,148 @@
+        // If the VO bits are available during tracing,
+        // we validate the objects we trace using the VO bits.
+        debug_assert!(
+            vo_bit::is_vo_bit_set::<VM>(object),
+            "{:x}: VO bit not set",
+            object
+        );
+    }
+}
+
+pub(crate) fn on_object_marked<VM: VMBinding>(object: ObjectReference) {
+    match strategy::<VM>() {
+        ImmixVOBitUpdateStrategy::ClearAndReconstruct => {
+            // In this strategy, we set the VO bit when an object is marked.
+            vo_bit::set_vo_bit::<VM>(object);
+        }
+        ImmixVOBitUpdateStrategy::CopyFromMarkBits => {
+            // VO bit was not cleared before tracing in this strategy.  Do nothing.
+        }
+    }
+}
+
+pub(crate) fn on_object_forwarded<VM: VMBinding>(new_object: ObjectReference) {
+    match strategy::<VM>() {
+        ImmixVOBitUpdateStrategy::ClearAndReconstruct => {
+            // In this strategy, we set the VO bit of the to-space object when forwarded.
+            vo_bit::set_vo_bit::<VM>(new_object);
+        }
+        ImmixVOBitUpdateStrategy::CopyFromMarkBits => {
+            // In this strategy, we will copy mark bits to VO bits.
+            // We need to set mark bits for to-space objects, too.
+            VM::VMObjectModel::LOCAL_MARK_BIT_SPEC.store_atomic::<VM, u8>(
+                new_object,
+                1,
+                None,
+                Ordering::SeqCst,
+            );
+
+            // In theory, we don't need to set the VO bit for to-space objects because we
+            // will copy the VO bits from mark bits during Release.  However, Some VMs
+            // allow the same edge to be traced twice, and MMTk will see the edge pointing
+            // to a to-space object when visiting the edge the second time.  Considering
+            // that we may want to use the VO bits to validate if the edge is valid, we set
+            // the VO bit for the to-space object, too.
+            vo_bit::set_vo_bit::<VM>(new_object);
+        }
+    }
+}
+
+pub(crate) fn on_block_swept<VM: VMBinding>(block: &Block, is_occupied: bool) {
+    match strategy::<VM>() {
+        ImmixVOBitUpdateStrategy::ClearAndReconstruct => {
+            // Do nothing.  The VO bit metadata is already reconstructed.
+        }
+        ImmixVOBitUpdateStrategy::CopyFromMarkBits => {
+            // In this strategy, we need to update the VO bits state after marking.
+            if is_occupied {
+                // If the block has live objects, copy the VO bits from mark bits.
+                vo_bit::bcopy_vo_bit_from_mark_bit::<VM>(block.start(), Block::BYTES);
+            } else {
+                // If the block has no live objects, simply clear the VO bits.
+                vo_bit::bzero_vo_bit(block.start(), Block::BYTES);
+            }
+        }
+    }
+}
+
+/// A work packet to clear VO bit metadata after Prepare.
+pub struct ClearVOBitsAfterPrepare {
```

Comment:
> Can't we run it in parallel with other prepare packets?

No.  One important user of the VO bits is conservative stack scanning which happens in the `Prepare` bucket.

> ... plan-specific buckets...

The `ClearVOBits` bucket (or the ClearAndReconstruct strategy) is applicable for plans that cannot easily clear VO bits for dead objects.  Currently it is just Immix, but new GC algorithms in the future may have similar traits.  I suspect region-based collectors, such as G1, may have similar traits.

> ... adding new general buckets...

As for adding buckets, previous experiments (while developing the VMRefClosure bucket) show that the time overhead of new buckets is negligible if no work packets are added to it.

Reply:
OK. Probably add some comments to `WorkBucketStage::ClearVOBits` to say that we need VO bits for stack scanning which happens in `Prepare` so we have to clear the vo bits after prepare. I think we may eventually want a separate bucket for stack scanning, and put vo bits packets to prepare, as the vo bits work fit into the category of 'preparing something'.

If I remember correctly, the cost of adding an empty bucket is ~1% GC time. It is negligible if we look at the total time, but it does have an overhead, even if it is a no-op.

Reply:
> ... I think we may eventually want a separate bucket for stack scanning, and put vo bits packets to prepare, as the vo bits work fit into the category of 'preparing something'.

Speaking of adding separate buckets, I am considering refining or reimplementing the following PRs: https://github.com/mmtk/mmtk-core/pull/706 and https://github.com/mmtk/mmtk-core/pull/705 because Ruby needs "black roots".  #706 It introduces a separate bucket for immovable roots.  Those objects pointed by immovable roots need to be traced before any updateable edges are traced which may cause objects to move.

If that happens, the `ClearVOBits` bucket should be added before tracing objects pointed by red/black roots too, because those objects (and the transitive closure of red roots) will also participate in "reconstructing the VO bits during tracing".

Overall, the buckets should be

```
StopMutators -> ScanRoots -> ClearVOBits -> ClosureImmovable -> Closure
```

> 
> If I remember correctly, the cost of adding an empty bucket is ~1% GC time. It is negligible if we look at the total time, but it does have an overhead, even if it is a no-op.

There is a difference between "an empty bucket" and "a bucket that contains a packet that does nothing".  The latter has greater cost than the former because workers will have to wake up again, only to find a no-op packet, and go to sleep again and let the coordinator open more buckets.  If the bucket is completely empty (without even a no-op packet), the current `GCWorkScheduler::update_buckets` method can open multiple buckets at the same time.  If we are not using the `ClearAndReconstruct` strategy (or not using the Immix plan or its variants), that bucket will be completely empty.